README.md
# Recruiter

<!-- TABLE OF CONTENTS -->
<details>
  <summary>Table of Contents</summary>
  <ol>
    <li><a href="#project-description">Project Description</a></li>
    <li><a href="#related-resources">Related Resources</a></li>
    <li><a href="#runtime-environment">Runtime Environment</a></li>
    <li><a href="#commands">Commands</a></li>
    <li><a href="#development-environment">Development Environment</a></li>
    <li><a href="#test-environment">Test Environment</a></li>
    <li><a href="#deployment-instructions">Deployment instructions</a></li>
  </ol>
</details>

<!-- ABOUT THE PROJECT -->

## Project Description

Recruiter is system that is designed to help Recruiters and Talent Sources to find new candidates. It is a Python Bootcamp Project. The project is just a REST API service with basic functionality.

[![GitHub license](https://img.shields.io/github/license/DevIhor/Recruiter)](https://github.com/DevIhor/Recruiter/blob/main/LICENSE)
[![Pending Pull-Requests](https://img.shields.io/github/issues-pr/DevIhor/Recruiter?style=flat-square)](https://github.com/DevIhor/Recruiter/pulls)
[![GitHub top language](https://img.shields.io/github/languages/top/DevIhor/Recruiter)](https://img.shields.io/github/languages/top/DevIhor/Recruiter)

## Related Resources

The following list contains information of what is used in this project.

[![Django](https://img.shields.io/badge/Django-4.1-green?style=for-the-badge)](https://docs.djangoproject.com/en/4.1/)
[![DRF](https://img.shields.io/badge/DRF-3.13.1-green?style=for-the-badge)](https://www.django-rest-framework.org/)
[![Celery](https://img.shields.io/badge/Celery-5.2.7-green?style=for-the-badge)](https://docs.celeryq.dev/en/stable/)
[![Redis](https://img.shields.io/badge/Redis-4.3.4-green?style=for-the-badge)](https://redis.io/docs/)
[![PostreSQL](https://img.shields.io/badge/PostreSQL-12.12-green?style=for-the-badge)](https://www.postgresql.org/docs/)

We are using Jira for Kanban board:

[Link to the board](https://coaxpythonbootcamp.atlassian.net/jira/software/projects/CPB/boards/1)

## Runtime Environment

Environmental data is set up via .ENV files, that should have the following info:

```env
DJANGO_SETTINGS_MODULE=config.settings
PROJECT_NAME=
SITE_URL=
SECRET_KEY=
ALLOWED_HOSTS=*
CORS_ALLOW_ALL_ORIGINS=YES
EMAIL_BACKEND=
DEFAULT_FROM_EMAIL=
REDIS_URL=
EMAIL_HOST_USER =
EMAIL_HOST_PASSWORD =
DEBUG_MODE=YES
```

## Commands

```bash
# Run to create migrations
python manage.py makemigrations

# Run apply migrations
python manage.py migrate

# Collect static into Storage
python manage.py collectstatic --no-input

# Create superuser
python manage.py createsuperuser

# To enter interactive Django shell
python manage.py shell

# To run a development server
python manage.py runserver

# To run a celery worker
celery -A config.celery worker --loginfo=info
```

## Development Environment

In order to test this project on your local machine, do the following:

- Clone this repo to your local machine using the following command:

```bash
git clone https://github.com/DevIhor/Recruiter.git
```

- Navigate into the folder and install the requirements by running the following command:

```bash
cd Recruiter && pip install -r requirements/dev.txt
```

- Make the migrations to prepare the database:

```bash
python manage.py migrate
```

- Run the server by the following command:

```bash
python manage.py runserver
```

The project should be available on your localhost, check the output of your command line for the details.

## Test environment

We are using [coverage.py](https://coverage.readthedocs.io/en/6.4.4/) for testing and coverage.
Run the tests by running:

```bash
coverage run -m pytest
coverage report
```

## Deployment instructions

Live server is available at [this link](http://65.109.3.1/swagger/).


CHANGELOG.md
# Changelog

The file contains changelog for the project.

## [0.1.1] - [Unreleased]

- Added EmailLetter [CPB-22](https://github.com/DevIhor/Recruiter/pull/22)
- Reset user's password [CPB-29](https://github.com/DevIhor/Recruiter/pull/27)
- Added Vacancy CRUD api endpoint [CPB-33](https://github.com/DevIhor/Recruiter/pull/26)
- Created events views, factories and tests [CPB-35](https://github.com/DevIhor/Recruiter/pull/24)
- Merged settings into a single file [CPB-37](https://github.com/DevIhor/Recruiter/pull/23)
- Updated readme [CPB-45](https://github.com/DevIhor/Recruiter/pull/25)
- Updated changelog [CPB-46]

## [0.1.0] - [Unreleased] - 07.09.2022

- Project skeleton was created [CPB-1](https://github.com/DevIhor/Recruiter/pull/1)
- Updated Docs [CPB-2](https://github.com/DevIhor/Recruiter/pull/9)
- Pre-commit hook with a linter was added [CPB-3](https://github.com/DevIhor/Recruiter/pull/4)
- Added tests coverage [CPB-4](https://github.com/DevIhor/Recruiter/pull/7)
- Custom User model was created [CPB-5](https://github.com/DevIhor/Recruiter/pull/2)
- Profile model was created [CPB-6](https://github.com/DevIhor/Recruiter/pull/11)
- Configured email [CPB-7](https://github.com/DevIhor/Recruiter/pull/8)
- CORS headers were configured [CPB-8](https://github.com/DevIhor/Recruiter/pull/3)
- Environment variables were submitted [CPB-9](https://github.com/DevIhor/Recruiter/pull/1)
- Added Static and Media Configurations [CPB-12](https://github.com/DevIhor/Recruiter/pull/6)
- Candidate model was created [CPB-14](https://github.com/DevIhor/Recruiter/pull/5)
- Created recruiter and reviewer groups [CPB-16](https://github.com/DevIhor/Recruiter/pull/20)
- Vacancy model was created [CPB-17](https://github.com/DevIhor/Recruiter/pull/10)
- Candidate-Vacancy relation was added [CPB-18](https://github.com/DevIhor/Recruiter/pull/17)
- Added Event model, added EventType model [CPB-19](https://github.com/DevIhor/Recruiter/pull/14)
- EmailTemplate model was created [CPB-21](https://github.com/DevIhor/Recruiter/pull/12)
- Setuped DRF and JWT [CPB-26](https://github.com/DevIhor/Recruiter/pull/13)
- Sign In API was created [CPB-27](https://github.com/DevIhor/Recruiter/pull/19)
- Sign Up API was created [CPB-28](https://github.com/DevIhor/Recruiter/pull/18)
- Divide requirements into dev and prod [CPB-36](https://github.com/DevIhor/Recruiter/pull/16)


backend/config/settings.py
import os
from datetime import timedelta
from pathlib import Path

import dj_database_url
from corsheaders.defaults import default_headers, default_methods
from dotenv import load_dotenv

load_dotenv()

# Build paths inside the project like this: BASE_DIR / 'subdir'.
BASE_DIR = Path(__file__).resolve().parent.parent.parent

PROJECT_NAME = os.environ.get("PROJECT_NAME")
SITE_URL = os.environ.get("SITE_URL")

# Quick-start development settings - unsuitable for production
# See https://docs.djangoproject.com/en/4.1/howto/deployment/checklist/

# SECURITY WARNING: keep the secret key used in production secret!
SECRET_KEY = os.environ.get("SECRET_KEY")

# SECURITY WARNING: don't run with debug turned on in production!
DEBUG = os.environ.get("DEBUG_MODE") == "YES"

ALLOWED_HOSTS = os.environ.get("ALLOWED_HOSTS").split(",")


# Application definition

INSTALLED_APPS = [
    # Default apps
    "django.contrib.admin",
    "django.contrib.auth",
    "django.contrib.contenttypes",
    "django.contrib.sessions",
    "django.contrib.messages",
    "django.contrib.staticfiles",
    # Installed apps
    "ckeditor",
    "corsheaders",
    "rest_framework",
    "django_filters",
    "django_rest_passwordreset",
    "drf_yasg",
    "import_export",
    "rest_framework_simplejwt",
    "phonenumber_field",
    "storages",
    "taggit",
    # Custom apps
    "apps.accounts",
    "apps.vacancies",
    "apps.candidates",
    "apps.emails",
    "apps.events",
    "apps.resume",
]

MIDDLEWARE = [
    "django.middleware.security.SecurityMiddleware",
    "django.contrib.sessions.middleware.SessionMiddleware",
    "corsheaders.middleware.CorsMiddleware",  # new
    "django.middleware.common.CommonMiddleware",
    "django.middleware.csrf.CsrfViewMiddleware",
    "django.contrib.auth.middleware.AuthenticationMiddleware",
    "django.contrib.messages.middleware.MessageMiddleware",
    "django.middleware.clickjacking.XFrameOptionsMiddleware",
    "whitenoise.middleware.WhiteNoiseMiddleware",  # new
]

ROOT_URLCONF = "config.urls"

TEMPLATES = [
    {
        "BACKEND": "django.template.backends.django.DjangoTemplates",
        "DIRS": [],
        "APP_DIRS": True,
        "OPTIONS": {
            "context_processors": [
                "django.template.context_processors.debug",
                "django.template.context_processors.request",
                "django.contrib.auth.context_processors.auth",
                "django.contrib.messages.context_processors.messages",
            ],
        },
    },
]

WSGI_APPLICATION = "config.wsgi.application"


# Database
# https://docs.djangoproject.com/en/4.1/ref/settings/#databases

if DEBUG:
    DATABASES = {
        "default": {
            "ENGINE": "django.db.backends.sqlite3",
            "NAME": BASE_DIR / "db.sqlite3",
        }
    }
else:
    DATABASES = {
        "default": dj_database_url.config(default=os.environ.get("DATABASE_URL")),
    }


# Password validation
# https://docs.djangoproject.com/en/4.1/ref/settings/#auth-password-validators

AUTH_PASSWORD_VALIDATORS = [
    {
        "NAME": "django.contrib.auth.password_validation.UserAttributeSimilarityValidator",
    },
    {
        "NAME": "django.contrib.auth.password_validation.MinimumLengthValidator",
    },
    {
        "NAME": "django.contrib.auth.password_validation.CommonPasswordValidator",
    },
    {
        "NAME": "django.contrib.auth.password_validation.NumericPasswordValidator",
    },
]

REST_FRAMEWORK = {
    "DEFAULT_RENDERER_CLASSES": (
        "rest_framework.renderers.JSONRenderer",
        "rest_framework.renderers.TemplateHTMLRenderer",
    ),
    "DEFAULT_THROTTLE_RATES": {
        "anon": "1/second",
        "user": "10/second",
    },
    "DEFAULT_AUTHENTICATION_CLASSES": (
        "rest_framework_simplejwt.authentication.JWTAuthentication",
    ),
    "TEST_REQUEST_DEFAULT_FORMAT": "json",
}

SIMPLE_JWT = {
    "ACCESS_TOKEN_LIFETIME": timedelta(days=14),
    "REFRESH_TOKEN_LIFETIME": timedelta(days=90),
    "AUTH_HEADER_TYPES": ("JWT",),
}

AUTH_USER_MODEL = "accounts.User"

# Internationalization
# https://docs.djangoproject.com/en/4.1/topics/i18n/

LANGUAGE_CODE = "en-us"

TIME_ZONE = "UTC"

USE_I18N = True

USE_TZ = True


# Static files (CSS, JavaScript, Images)
# https://docs.djangoproject.com/en/4.1/howto/static-files/

STATIC_URL = "static/"
STATIC_ROOT = os.path.join(BASE_DIR, "staticfiles")
# STATICFILES_STORAGE = "whitenoise.storage.CompressedManifestStaticFilesStorage"

MEDIA_URL = "media/"

USING_AWS_S3_BUCKET_FOR_STORAGE = os.environ.get("USING_AWS_S3_BUCKET_FOR_STORAGE") == "YES"

# Here, the storage location of the media files is determined.
if DEBUG:
    MEDIA_ROOT = os.path.join(BASE_DIR, "mediafiles")
elif USING_AWS_S3_BUCKET_FOR_STORAGE:
    AWS_ACCESS_KEY_ID = os.environ.get("AWS_ACCESS_KEY_ID")
    AWS_SECRET_ACCESS_KEY = os.environ.get("AWS_SECRET_ACCESS_KEY")
    AWS_STORAGE_BUCKET_NAME = os.environ.get("AWS_STORAGE_BUCKET_NAME")

    AWS_S3_FILE_OVERWRITE = False  # so that files with the same name are not overwritten.
    AWS_DEFAULT_ACL = None  # file will be private per Amazon’s default.

    DEFAULT_FILE_STORAGE = "storages.backends.s3boto3.S3Boto3Storage"
else:
    # For using media on the server, additional server configuration is required.
    MEDIA_ROOT = os.path.join(BASE_DIR, "mediafiles")

# Default primary key field type
# https://docs.djangoproject.com/en/4.1/ref/settings/#default-auto-field

DEFAULT_AUTO_FIELD = "django.db.models.BigAutoField"


# List of sites from which you can make requests to this project.
# To enable CORS_ALLOWED_ORIGINS, change CORS_ALLOW_ALL_ORIGINS to False
CORS_ALLOW_ALL_ORIGINS = os.environ.get("CORS_ALLOW_ALL_ORIGINS") == "YES"

if not CORS_ALLOW_ALL_ORIGINS:
    CORS_ALLOWED_ORIGINS = os.environ.get("CORS_ALLOWED_ORIGINS").split(",")

# Available HTTP methods and headers that are allowed for the actual request.
CORS_ALLOW_METHODS = list(default_methods) + []
CORS_ALLOW_HEADERS = list(default_headers) + []


TAGGIT_CASE_INSENSITIVE = True

# Email
EMAIL_BACKEND = os.environ.get("EMAIL_BACKEND")
EMAIL_FILE_PATH = os.path.join(BASE_DIR, "emails")
DEFAULT_FROM_EMAIL = os.environ.get("DEFAULT_FROM_EMAIL")
EMAIL_USE_TLS = os.environ.get("EMAIL_USE_TLS") == "YES"
EMAIL_HOST = os.environ.get("EMAIL_HOST")
EMAIL_PORT = os.environ.get("EMAIL_PORT")
EMAIL_HOST_USER = os.environ.get("EMAIL_HOST_USER")
EMAIL_HOST_PASSWORD = os.environ.get("EMAIL_HOST_PASSWORD")

# Celery
CELERY_BROKER_URL = os.environ.get("REDIS_URL")
CELERY_RESULT_BACKEND = os.environ.get("REDIS_URL")
CELERY_TASK_TIME_LIMIT = 5 * 60

# SWAGGER
if DEBUG:
    SWAGGER_SETTINGS = {
        "exclude_namespaces": [],
        "USE_SESSION_AUTH": False,
        "PERSIST_AUTH": True,
        "SECURITY_DEFINITIONS": {
            "Bearer": {"type": "apiKey", "name": "Authorization", "in": "header"}
        },
        "SUPPORTED_SUBMIT_METHODS": ["get", "put", "post", "delete", "patch"],
        "SHOW_REQUEST_HEADERS": True,
    }

REDOC_SETTINGS = {
    "LAZY_RENDERING": False,
}
